### ISSUE TITLE
DBB Jenkins integration issue

### ISSUE THREAD

**Issue by nk2164** [IBM/dbb#31]
I had set up the jenkins integration on my jenkins 2.60.3 many months back and all of this worked great and i am able to successfully build the mortgage application.

https://www.ibm.com/support/knowledgecenter/SS6T76_1.0.2/jenkinsintegration.html

Old set up
Jenkins : 2.60.3
SSH Slave Plugin : 1.26
Git client plugin : 2.7.3
Git Plugin : 3.9.1
Groovy Plugin: 2.0

We recently updated jenkins to a much higher 2.222.1. This is our new set up. Now the build does not work.I can see it runs the git-jenkins.sh rev-parse . Then i see a FATAL:Invalid id and the git command shows junk chars. Below is the plugin versions on my new server.

Jenkins : 2.222.1
SSH Build Agent Plugin : 1.31.2
Git Plugin : 4.2.2
Git Client Plugin: 3.2.1
Groovy Plugin: 2.2

Have you been able to get this mortgage app build to run on a newer version of jenkins ? 

I also see warning in the log when creating the new mainframe agent . I use the ssh build agent plugin to connect on my new jenkins server. I see the agent successfully connected message at the end of log ...but before that there is a warning 'linkage error while performing user request' java.lang.Unsatisfiedlinkerror: Native library (com/sun/jna/z/os-390x/libjnidispatch.so) not found in resource path. There is also an error in the log "ERROR" Failed to monitor free swap space'.

The new jenkins no longer puts a slave.jar but instead a remoting.jar on the agent machine. 

Please let me know if there are any changes needed when trying to run with a more latest jenkins version

**Comment by dennis-behm** [comment]
Hi nk2164,

do you use a ssh passphrase or http credentials managed by Jenkins? Does the build lists the remote branches on origin?

The Git client plugin has changed its behaviour in this area - which causes an encoding issue with the credentials, leading to a `FATAL:Invalid Id`. We are currently working on a new git-jenkins.sh.

It is currently a bank holiday over here, I will create a PR tomorrow, so you can see the updated version of git-jenkins.sh

**Comment by nk2164** [comment]
Hi Dennis,

I don't use a passphrase. Here is my set up

Agent set up
Credentials - application id/password stored in jenkins
Host verification strategy - "Manually provided key verification strategy" and the ssh key for the host.
In the node properties, i updated the tool location to point git path to the git-jenkins.sh file on remote USS

In the jenkins build job for mortgage application, i chose "git" in source code management tab and then provide the ssh url (git@xx.xx.xx.org/repo-name.git) and for credential i use the application id and private key for that id . This id has access to the git repository . This id is stored in jenkins.

In the jenkins log , i see it first tries to run rev-parse --is-inside-work-tree # timeout=10

Fetching changes from the remote git repository

git-jenkins.sh config remote.origin.url git@xx.xx.xx.org/repo-name.git # timeout = 10

Fetching upstream changes from git@xx.xx.xx.org/repo-name.git

The multiple lines of 
> /dbbz/bin/git-jenkins.sh --version # timeout=10
> /dbbz/bin/git-jenkins.sh --version # timeout=10
> /dbbz/bin/git-jenkins.sh --version # timeout=10
> /dbbz/bin/git-jenkins.sh --version # timeout=10

Using GIT_SSH to set credentials to my  (user id and private key with access to git repo)

After that , i see 2 entries for following
> /dbbz/bin/git-jenkins.sh fetch --tags git@xx.xx.xx.org/repo-name.git +refs/heads/*:refs/remotes/origin/*  # timeout=10

> /dbbz/bin/git-jenkins.sh rev-parse origin/master^{commit} #timeout = 10

FATAL:INVALID Id: fatal : ambiguous argument 'origin/master^{commit}' : unknown revision or path not in the working tree
use '--' to separate paths from revisions, like this : 
'git <command> [<revision>...] -- [<file>...]'
@(*##)@__###_ (some junk chars)

Then i see a stack trace...

**Comment by dennis-behm** [comment]
Ok, you are managing your private key in Jenkins, no ssh passphrase.
> In the jenkins build job for mortgage application, i chose "git" in source code management tab and then provide the ssh url (git@xx.xx.xx.org/repo-name.git) and for credential i use the application id and private key for that id . This id has access to the git repository . This id is stored in jenkins.

Before going for the new script can you please check the two options:
- just run git directly and don't go through the git-jenkins.sh wrapper (you can configure that at the agent)
- Instead of the iconv at the very end, just invoke git without the magic grep:
```
#git "$@" | iconv -f ibm-1047 -t ibm-1047
git "$@"
```

**Comment by nk2164** [comment]
Hi Dennis, The first option works for me. I changed the agent to point directly to the rocket git path /usr/local/bin/rocket/bin/git and the job is building successfully. Thank you for your help.

I also tried the second option of removing the iconv and just having `git "$@"`  
This time , after the following command 
`fetch --tags --progress git@xx.xx.org:xx/repo.git +refs/heads/*:refs/origin/* # timeout=10`
I see a 
```
ERROR:Error cloning remote repo 'origin'
hudson.plugin.git.exception command "fetch --tags --progress git@xx.xx.org:xx/repo.git +refs/heads/*:refs/origin/*" return status code is 128"
stdout : --- a lot of junk  chars---

```
==============
for the USS agent log i see the successfully connected message but also this warning at the end . My DBB builds are running fine though (with the change you mentioned in first option to directly use git).  The logs on my old jenkins does not have these warnings and errors .

```
at com.sun.jna.Native.loadNativeDispatchLibraryFromClasspath(Native.java:962)
at com.sun.jna.Native.loadNativeDispatchLibrary(Native.java:922)
at com.sun.jna.Native.<clinit>(Native.java:190)
at hudson.util.jna.GNUCLibrary.<clinit>(GNUCLibrary.java:115)
at jenkins.slaves.StandardOutputSwapper$ChannelSwapper.swap(StandardOutputSwapper.java:60)
at jenkins.slaves.StandardOutputSwapper$ChannelSwapper.call(StandardOutputSwapper.java:45)
at jenkins.slaves.StandardOutputSwapper$ChannelSwapper.call(StandardOutputSwapper.java:39)
at hudson.remoting.UserRequest.perform(UserRequest.java:212)
at hudson.remoting.UserRequest.perform(UserRequest.java:54)
at hudson.remoting.Request$2.run(Request.java:369)
at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
at java.util.concurrent.FutureTask.run(FutureTask.java:277)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
at java.lang.Thread.run(Thread.java:811)

```
This is followed by these erorrs and stack traces

```
Error : Failed to monitor Free Temp Space
java.util.concurrent.TimeoutException
Error : Failed to monitor Free Disk Space
java.util.concurrent.TimeoutException
Error : Failed to monitor Free Swap Space
java.util.concurrent.TimeoutException
```

**Comment by dennis-behm** [comment]
Hi nk2164, 
It would be interesting to see what these 
> --- a lot of junk  chars---

are ... Can you Copy/Paste them to a USS file and play a bit with the encoding using IDZ or zExplorer or iconv?

Keep in mind, using git directly will cause an issue as soon another applications decide to use a ssh passphrase or start using http with credentials.

Re. your agent. I see the same missing object `java.lang.UnsatisfiedLinkError: Native library (com/sun/jna/z/os-s390x/libjnidispatch.so) not found in resource path` and consider this to be normal, because we run the IBM java runtime and not Oracle.

**Comment by nk2164** [comment]
Hi Dennis,

I tried copying into a file in USS and tried changing the encoding and it did not help. Is there a way i can capture the output of the git command and route that to a file on USS ? I will try some more tomorrow.

I also tried the mortgage build with your updated file and that works as well.

So based on this below comment from you , I'm thinking its best to start using the updated git wrapper right ?

> Keep in mind, using git directly will cause an issue as soon another applications decide to use a ssh passphrase or start using http with credentials.
